presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -145,26 +159,28 @@ private void startLoadSplits()
             return;
         }
         if (outstandingTasks.incrementAndGet() > maxPartitionBatchSize) {
-            outstandingTasks.decrementAndGet();
+            if (outstandingTasks.decrementAndGet() == 0) {
+                invokeFinishedIfShould();
+            }
             return;
         }
         executor.execute(() -> {
             try {
-                loadSplits();
+                threeQueuesRwLock.readLock().lock();
+                try {
+                    loadSplits();
+                }
+                finally {
+                    threeQueuesRwLock.readLock().unlock();
+                }
                 if (!hiveSplitSource.isQueueFull()) {
```

Comment:
I think there's also a race with `resume()` that could result in the loader no longer making progress. Suppose there's one thread, A, and it finds that `hiveSplitSource` is full, so it moves to line 182, but before it decrements `outstandingTasks` the `hiveSplitSource` is emptied and `resume()` is called. Since `resume()` has a `outstandingTasks == 0` check it will be a no-op. Then thread A will terminate, and we will have missed the `resume()` signal